Code/Tools/TestImpactFramework/Runtime/Common/Code/Include/Static/BuildTarget/Common/TestImpactBuildGraph.h

```diff
@@ -0,0 +1,174 @@
+    struct BuildGraphVertex;
+    
+    //! Build graph target set for dependencies or dependers.
+    template<typename ProductionTarget, typename TestTarget>
+    using TargetBuildGraphSet = AZStd::unordered_set<const BuildGraphVertex<ProductionTarget, TestTarget>*>;
+    
+    //! Build graph for the dependencies and dependers of a given build target.
+    template<typename ProductionTarget, typename TestTarget>
+    struct TargetBuildGraph
+    {
+        TargetBuildGraphSet<ProductionTarget, TestTarget> m_build; //!< Build dependencies/dependers.
+        TargetBuildGraphSet<ProductionTarget, TestTarget> m_runtime; //!< Runtime dependencies/dependers.
+    };
+    
+    //! Vertex in the build graph
+    template<typename ProductionTarget, typename TestTarget>
+    struct BuildGraphVertex
+    {
+        BuildGraphVertex(BuildTarget<ProductionTarget, TestTarget> buildTarget)
+            : m_buildTarget(buildTarget)
+        {
+        }
+
+        BuildTarget<ProductionTarget, TestTarget> m_buildTarget; //!< The build target for this vertex.
+        TargetBuildGraph<ProductionTarget, TestTarget> m_dependencies; //!< The dependencies of this build target.
+        TargetBuildGraph<ProductionTarget, TestTarget> m_dependers; //!< The dependers of this build target.
+    };
+    
+    //! Build graph of all build targets in the repository, including their depenency and depender graphs.
+    template<typename ProductionTarget, typename TestTarget>
+    class BuildGraph
+    {
+    public:
+        BuildGraph(const BuildTargetList<ProductionTarget, TestTarget>& buildTargetList);
+    
+        //! Returns the vertex for the specified build target, else returns `nullptr`.
+        const BuildGraphVertex<ProductionTarget, TestTarget>* GetVertex(
+            const BuildTarget<ProductionTarget, TestTarget>& buildTarget) const;
+
+        //! Returns the vertex for the specified build target, else throws `BuildTargetException`.
+        const BuildGraphVertex<ProductionTarget, TestTarget>& GetVertexOrThrow(
+            const BuildTarget<ProductionTarget, TestTarget>& buildTarget) const;
+            
+    private:
+        //! Map of all graph vertices, identified by build target.
+        AZStd::unordered_map<BuildTarget<ProductionTarget, TestTarget>, BuildGraphVertex<ProductionTarget, TestTarget>>
+            m_buildGraphVertices;
+    };
+    
+    template<typename ProductionTarget, typename TestTarget>
+    BuildGraph<ProductionTarget, TestTarget>::BuildGraph(
+        const BuildTargetList<ProductionTarget, TestTarget>& buildTargetList)
+    {
+        // Build dependency graphs
+        for (const auto& buildTarget : buildTargetList.GetBuildTargets())
+        {
+            const auto addOrRetrieveVertex = [this](const BuildTarget<ProductionTarget, TestTarget>& buildTarget)
+            {
+                auto it = m_buildGraphVertices.find(buildTarget);
+                if (it == m_buildGraphVertices.end())
```

Comment:
nit: It's possible to move `it` to an if initializer (it's available in the else too)

Reply:
Fixed 👍